src/ol/resolutionconstraint.js

```diff
@@ -44,29 +83,74 @@ export function createSnapToResolutions(resolutions) {
 /**
  * @param {number} power Power.
  * @param {number} maxResolution Maximum resolution.
- * @param {number=} opt_maxLevel Maximum level.
+ * @param {number=} opt_minResolution Minimum resolution.
+ * @param {boolean=} opt_smooth If true, the view will be able to slightly exceed resolution limits. Default: true.
+ * @param {import("./extent.js").Extent=} opt_maxExtent Maximum allowed extent.
  * @return {Type} Zoom function.
  */
-export function createSnapToPower(power, maxResolution, opt_maxLevel) {
+export function createSnapToPower(power, maxResolution, opt_minResolution, opt_smooth, opt_maxExtent) {
   return (
     /**
      * @param {number|undefined} resolution Resolution.
-     * @param {number} delta Delta.
      * @param {number} direction Direction.
+     * @param {import("./size.js").Size} size Viewport size.
+     * @param {boolean=} opt_isMoving True if an interaction or animation is in progress.
      * @return {number|undefined} Resolution.
      */
-    function(resolution, delta, direction) {
+    function(resolution, direction, size, opt_isMoving) {
       if (resolution !== undefined) {
-        const offset = -direction / 2 + 0.5;
-        const oldLevel = Math.floor(
-          Math.log(maxResolution / resolution) / Math.log(power) + offset);
-        let newLevel = Math.max(oldLevel + delta, 0);
-        if (opt_maxLevel !== undefined) {
-          newLevel = Math.min(newLevel, opt_maxLevel);
+        const cappedMaxRes = opt_maxExtent ? getCappedResolution(maxResolution, opt_maxExtent, size) : maxResolution;
+        const minResolution = opt_minResolution !== undefined ? opt_minResolution : 0;
+
+        // during interacting or animating, allow intermediary values
+        if (opt_isMoving) {
+          const smooth = opt_smooth !== undefined ? opt_smooth : true;
+          if (!smooth) {
+            return clamp(resolution, minResolution, cappedMaxRes);
+          }
+          return getSmoothedResolution(resolution, cappedMaxRes, minResolution);
         }
-        return maxResolution / Math.pow(power, newLevel);
+
+        const offset = -direction * (0.5 - 1e-9) + 0.5;
```

Comment:
What exactly does this calculation do? I'm particularly wondering about `1e-9`.

Reply:
This is needed in the case where the direction is -1: the offset applied to the zoom level should be just below 1, otherwise even on a round value the zoom level will be increased at the following line:
https://github.com/openlayers/openlayers/blob/7ca32cd40b0d3169e8b8a429446088939d9b5aca/src/ol/resolutionconstraint.js#L116-L117